### PR TITLE
Implement remaining functionality for Project framework

### PR DESCRIPTION
**Pull request by chia-yh**
`Project` class lacks functionality to store in .`json` save file, restore from save file. Placeholder `VALIDATION_REGEX` for some `Project` fields.

Function to store in `.json` file is more user-friendly as it allows users to avoid having to create instances of `Project` every time `SOCket` is launched. `VALIDATION_REGEX` helps ensure fields entered are valid

Let's,
* update `SOCket` to allow `Project` instances to be stored in the `.json` format
* update the storage module to include `Project` instances when saving to/reading from `.json` save files
* add/update the relevant tests to ensure storage functionality is working as intended

**Design Considerations**
* storage in the `.json` format does not provide much support for references
* current implementation of `Project` storage has each instance write their own copies of the associated `Person` instances, "projects" is a separate attribute representing a separate list in the `.json` file, when reading from the `.json` file, the `toModelType()` method restores the references from `Project` instances to `Person` instances by matching the instances in the "persons" attribute list; if no matching `Person` instance is found, an `IllegalValueException` is thrown, indicating the lack of a matching `Person` instance under "persons
* having bidirectional navigability between `Person` and `Project` instances was attempted, however the implementation ran into big issues, one of which being the need to have fields no longer be immutable in order to set the bidirectional reference (no `final`), which, on top of the trouble implementing the restoration of `Project` > `Person` references, led me to scrap the idea

**Changes**
`ReadOnlySocket.java`:
* add getProjectList() method to return unmodifiable view of project list

`Socket.java`:
* add `UniqueProjectList` field to handle `Project` instances
* add `setProjects()` method to replace contents of `projects` list
* add project-level operations, `hasProject()`, `addProject()`, `setProject()`, `removeProject()`
* update `toString()` util method
* override `getProjectList()` method
* update `equals()` method to consider `projects`

`UniquePersonList.java`:
* add `getPerson()` method to facilitate sync of reference in `Project` instances when restoring from `.json` file

`ProjectDeadline.java`:
* remove `VALIDATION_REGEX`, `isValidProjectDeadline` utilises `DateTimeParseException` instead

`ProjectRepoHost.java`:
* update comments for references for validation
* update `MESSAGE_CONSTRAINTS`

`ProjectRepoName.java`:
* update comments for references for validation
* update `MESSAGE_CONSTRAINTS`

`SampleDataUtil.java`:
* add JavaDoc comments
* add `getSampleProjects()` method to return sample `Project` instances, without references to `Person` instances
* update `getSampleSocket()` to include sample `Project` instances
* add `getProjectSet()` method to return `Set<Project>`

`JsonAdaptedPerson.java`:
* fix spelling error

`JsonAdaptedProject.java`:
* add functionality to facilitate storing/reading of `Project` instances in `.json` format

`JsonSerializableSocket.java`:
* add `String` constants to indicate new error messages
* add `List<JsonAdaptedProject>` field to handle `Project` instances in `SOCket`
* update `toModelType()` method to match `Person` instances in `Project` instances with matching instances in `projects`, and
replace instances with references

`JsonSocketStorage.java`:
* rename remaining AddressBook references

`duplicatePersonSocket.json`:
* update with empty "projects" attribute

`duplicateProjectSocket.json`:
* populate with sample data for testing

`invalidPersonSocket.json`:
* update with empty "projects" attribute

`invalidProjectSocket.json`:
* populate with sample data for testing

`missingProjectMemberInPersonsSocket.json`:
* populate with sample data for testing

`typicalPersonsSocket.json`:
* update with empty "projects" attribute

`typicalProjectsSocket.json`:
* populate with sample data for testing

`SocketTest.java`:
* update `SocketStub`

`ProjectRepoHostTest.java`:
* update comments to reflect proper naming

`ProjectRepoNameTest.java`:
* update comments to reflect proper naming
* update tests to test correct validation behavior as per updated `VALIDATION_REGEX`

`JsonAdaptedProjectTest.java`:
* add tests to confirm behavior of `JsonAdaptedProject` instances

`JsonSerializableSocketTest.java`:
* add fields to store `Path` instances to added `.json` test files
* add tests to test `.json` storage of `Project` instances

`TypicalProjects.java`:
* add method to return `Socket` instance populated with data from `TypicalProjects`

### CODE REVIEW

File: src/main/java/seedu/socket/model/project/ProjectDeadline.java [hunk 1]
```diff
@@ -39,10 +51,11 @@ private Boolean isValidProjectDeadline(String test) {
      */
     public LocalDateTime toLocalDateTime() {
         if (deadline == "") {
-            return null;
+            return null; // this is dangerous TODO consider rethinking implementation
```
Comment (dillongoh) [inline src/main/java/seedu/socket/model/project/ProjectDeadline.java:54]: Can consider not having this method at all and only parsing the string if and when a LocalDateTime object is required. Alternatively, can use an assertion or throw an exception to ensure this method only returns when the deadline string is not empty.

**Review by dillongoh — APPROVED** [review]
LGTM! Can merge after deciding on whether the toLocalDateTime() implementation needs to be changed.

File: src/main/java/seedu/socket/model/project/ProjectDeadline.java (hunk 1 above)
Comment (chia-yh) [reply src/main/java/seedu/socket/model/project/ProjectDeadline.java:54]: In favor of assertion or exceptions
@cmang12 @huanghao1998 @nus-jeff 
thoughts?

Comment (chia-yh) [reply src/main/java/seedu/socket/model/project/ProjectDeadline.java]: Modified to throw DateTimeParseException for now, do feel free to suggest alternative implementations, perhaps when the use of this method is clearer

**Review by dillongoh — APPROVED** [review]
LGTM 👍

**Pull request merged**